### PR TITLE
Add warning when set "alias" in low level annontated field

### LINKED ISSUE #7828: Warn when `Field(alias=...)` is used as nested `Annotated` metadata
### Initial Checks

- [X] I confirm that I'm using Pydantic V2

### Description

Consider the following models:

```python
class AnnotatedOptional(BaseModel):
    field: Annotated[Optional[int], Field(alias='field_alias', gt=0)]


class OptionalAnnotated(BaseModel):
    field: Optional[Annotated[int, Field(alias='field_alias', gt=0)]]
```

The only difference is the order of `Optional` and `Annotated`.

`AnnotatedOptional` works as expected:

```python
# OK
AnnotatedOptional(field_alias=1)
# OK
AnnotatedOptional(field_alias=None)
# ValidationError: Input should be greater than 0
AnnotatedOptional(field_alias=-1)
```

On the other hand, `OptionalAnnotated` ignores `alias`:

```python
# The alias is ignored: ValidationError: Field required
OptionalAnnotated(field_alias=1)
# OK if the field name is used
OptionalAnnotated(field=1)
```

Strangely enough, _some_ `Field` parameters are respected, e.g., `gt` works as expected:

```python
# ValidationError: Input should be greater than 0
OptionalAnnotated(field=-1)
```

### Example Code

_No response_

### Python, Pydantic & OS Version

```Text
pydantic version: 2.4.2
        pydantic-core version: 2.10.1
          pydantic-core build: profile=release pgo=true
                 install path: [redacted]/virtualenvs/pydantic-optional-annotated-bug/lib/python3.12/site-packages/pydantic
               python version: 3.12.0 (main, Oct  5 2023, 13:09:34) [GCC 11.4.0]
                     platform: Linux-5.15.0-86-generic-x86_64-with-glibc2.35
             related packages: typing_extensions-4.8.0
```

### PR DESCRIPTION
**Pull request by nix010**
## Change Summary

Add warning message when user setup annotated field with `alias` kwarg at low level. #7828.

## Checklist

* [x] The pull request title is a good summary of the changes - it will be used in the changelog
* [x] Unit tests for the changes exist
* [x] Tests pass on CI
* [x] Documentation reflects the changes where applicable
* [x] My PR is ready to review, **please add a comment including the phrase "please review" to assign reviewers**

Fix #7828

Selected Reviewer: @adriangb

### CODE REVIEW

**Comment by nix010** [comment]
please review.

File: pydantic/_internal/_fields.py
Comment (dmontagu) [inline pydantic/_internal/_fields.py:239]:
```suggestion
            for anno_arg in getattr(ann_type, '__args__', ()):
                if not _typing_extra.is_annotated(anno_arg):
                    continue
                for anno_type_arg in _typing_extra.get_args(anno_arg):
                    if isinstance(anno_type_arg, FieldInfo) and anno_type_arg.alias is not None:
                        warnings.warn(
                            f'Field "{ann_name}" has `alias` should be set at higher level to have affect',
                            UserWarning,
                        )
                        break
```
Note: I removed the second break, and it seems to me like including it is problematic — if there are multiple annotated args it looks to me like it won't hit them.

I would prefer if we broke this out into a separate function (private in this module) to get this check moved out-of-line instead of making this method bigger and more complicated than it already is. (In that case, the `break` can just become `return`)

Comment (nix010) [reply pydantic/_internal/_fields.py:239]: @dmontagu Thank for the suggestions.
As my understand and please correct me if i'm wrong, warning on the first annotated args that has `alias` set or on all annotated args that have it is the same things !?.

Comment (sydney-runkle) [reply pydantic/_internal/_fields.py:239]: I think we want to warn for each argument where it's a problem, which is what @dmontagu is pointing out.

I agree with his suggestions above! Feel free to ping me when you've made the changes, and I can do a second review!

Comment (sydney-runkle) [reply pydantic/_internal/_fields.py:239]: @nix010,

Let me know if you have any questions about this suggestion. Thanks for your work on this!

Comment (nix010) [reply pydantic/_internal/_fields.py]: @sydney-runkle Made the changes, please review. Lost my laptop in a fire accident so it took a bit longer than usual.

File: pydantic/_internal/_fields.py
Comment (sydney-runkle) [inline pydantic/_internal/_fields.py:255]:
```suggestion
def _warn_on_nested_alias_in_annotation(ann_type: type[Any], ann_name: str):
```

File: pydantic/_internal/_fields.py
Comment (sydney-runkle) [inline pydantic/_internal/_fields.py:229]:
```suggestion
            _warn_on_nested_alias_in_annotation(ann_type, ann_name)
```

File: pydantic/_internal/_fields.py
Comment (sydney-runkle) [inline pydantic/_internal/_fields.py:264]:
```suggestion
                            f'`alias` specification on field "{ann_name}" must be set on outermost annotation to take effect.',
```

File: tests/test_annotated.py
Comment (sydney-runkle) [inline tests/test_annotated.py:237]:
```suggestion
        UserWarning, match=r'`alias` specification on field "low_level_alias_field" must be set on outermost annotation to take effect.
```

File: pydantic/_internal/_fields.py
Comment (sydney-runkle) [reply pydantic/_internal/_fields.py]: Resolving as fixed.

File: tests/test_annotated.py
Comment (sydney-runkle) [inline tests/test_annotated.py:237]:
```suggestion
        UserWarning, match=r'`alias` specification on field "low_level_alias_field" must be set on outermost annotation to take effect.'
```

**Comment by sydney-runkle** [comment]
@nix010,

Great work here, thanks! Will merge once tests pass - I made a few minor modifications to function names + warning wording.

File: tests/test_annotated.py
Comment (sydney-runkle) [inline tests/test_annotated.py:237]:
```suggestion
        UserWarning, 
        match=r'`alias` specification on field "low_level_alias_field" must be set on outermost annotation to take effect.'
```

File: tests/test_annotated.py
Comment (sydney-runkle) [inline tests/test_annotated.py:239]:
```suggestion
    with pytest.warns(
        UserWarning,
        match=r'`alias` specification on field "low_level_alias_field" must be set on outermost annotation to take effect.'
    ):
```

Comment (sydney-runkle) [reply tests/test_annotated.py:239]: Fixing the other linting issue I introduced

**Comment by nix010** [comment]
@sydney-runkle resolved the linting. Should be good now.

**Review by sydney-runkle — APPROVED** [review]
Great work here, thanks!

**Pull request merged**
